de.dlr.sc.virsat.model.extension.mechanical/src/de/dlr/sc/virsat/model/extension/mechanical/catia/CatiaExporter.java

```diff
@@ -117,41 +120,61 @@ public JsonObject transformElement(IBeanStructuralElementInstance element) {
 		return jsonElement;
 	}
 
+	
 	/**
 	 * Creates a JSON product object 
 	 * @param productBean bean to get visualisation values from
 	 * @return JSON product object
 	 */
 	private JsonObject transformProduct(IBeanStructuralElementInstance productBean) {
-		Visualisation vis = productBean.getFirst(Visualisation.class);
-
-		JsonObject jsonProduct = transformElement(productBean);
-		
-		if (vis != null) {
-			fillProductVisualisationFields(vis, jsonProduct);
-		} else if (hasDeepVisualisation(productBean)) {
-			fillProductDummyVisualisationFields(jsonProduct);
-		} else {
-			return null;
-		}
-		
-		IBeanStructuralElementInstance part = getPartForProduct(productBean);
-		parts.add(part);
+		mapBeansToJsonProducts = new HashMap<>();
+		BeanStructuralTreeTraverser traverser = new BeanStructuralTreeTraverser();
 		
-		jsonProduct.put(CatiaProperties.PRODUCT_ED_UUID.getKey(), part.getUuid());
-		jsonProduct.put(CatiaProperties.PRODUCT_REFERENCE_NAME.getKey(), part.getName());
-		
-		List<IBeanStructuralElementInstance> subProducts = productBean.getChildren(IBeanStructuralElementInstance.class);
-		JsonArray children = new JsonArray();
-		for (IBeanStructuralElementInstance subProduct: subProducts) {
-			JsonObject jsonSubProduct = transformProduct(subProduct);
-			if (jsonSubProduct != null) {
-				children.add(jsonSubProduct);
+		traverser.traverse(productBean, new IBeanStructuralTreeTraverserMatcher() {
+			Visualisation visCa;
+			boolean hasVisCa;
+			boolean isRoot;
+			
+			@Override
+			public boolean isMatching(IBeanStructuralElementInstance treeNode) {
+				visCa = treeNode.getFirst(Visualisation.class);
+				hasVisCa = visCa != null;
+				isRoot = treeNode.isRoot();

```

Comment:
Should we maybe check `treeNode == productBean` for root? This would correctly handle exporting a subtree.